entitybroker/core-providers/src/java/org/sakaiproject/entitybroker/providers/MembershipEntityProvider.java

```diff
@@ -309,6 +316,113 @@ public ActionReturn handleSiteMemberships(EntityView view, Map<String, Object> p
         return getEntities(new EntityReference(PREFIX, ""), s);
         
     }
+
+    /**
+     * Special handler for JSON uploads of site membership data. Takes a parameter 'json' which
+     * should contain json of the form:
+     * 
+     * [ {"id": "user1", "role": "access"},{"id": "user2","role": "maintain"} ]
+     */
+    @EntityCustomAction(action = "sitebyjson", viewKey = EntityView.VIEW_NEW)
+    public ActionReturn handleSiteJsonUpload(EntityView view, Map<String, Object> params) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("handleSiteJsonUpload method=" + view.getMethod() + ", params=" + params);
+        }
+
+        String siteId = view.getPathSegment(2);
+        if (siteId == null) {
+            siteId = (String) params.get("siteId");
+            if (siteId == null) {
+                throw new IllegalArgumentException(
+                        "siteId must be set in order to get site memberships, set in params or in the URL /membership/site/siteId");
+            }
+        }
+        
+        String locationReference = "/site/" + siteId;
+        
+        String json = (String) params.get("json");
+        if (json == null) {
+        	throw new IllegalArgumentException(
+                        "The membership JSON data must be supplied as a POST parameter named 'json'.");
+        }
+        
+        Map<String, String> extraResponseHeaders = new HashMap<String, String>(3);
+        
+        ObjectMapper mapper = new ObjectMapper();
+        List<JsonUser> memberships;
+		
+        try {
+            memberships = mapper.readValue(json.getBytes(), new TypeReference<List<JsonUser>>() { });
+        } catch(JsonParseException jpe) {
+            throw new IllegalArgumentException("The supplied JSON was invalid. Have a look at http://www.json.org/.");
+        } catch(JsonMappingException jpe) {
+            throw new IllegalArgumentException("The supplied JSON was invalid. Take a look at /direct/membership/describe for the correct structure.");
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Failed to read the supplied JSON.");
+        }
+		
+        // Collect the users into roles
+        Map<String, List<String>> usersToRoleMap = new HashMap<String, List<String>>();
+        for (JsonUser user : memberships) {
+            String id = user.getId();
+            String role = user.getRole();
+            
+            if (usersToRoleMap.containsKey(role)) {
+                usersToRoleMap.get(role).add(id);
+            } else {
+                List<String> users = new ArrayList<String>(20); // Prealloc space for 20 users in each role
```

Comment:
Not an issue, but this doesn't really save a lot in terms of reallocation costs as we can't really guess how big the list of users is going to be: could be 20, could be 2000. Might even be just 1 or 2 in which case this is wasting (a tiny bit of) memory.